### PR TITLE
adding HF metrics

### PR DESCRIPTION
**Pull request by liususan091219**
Adding HF metrics to sklearn_metric_loss_score

### CODE REVIEW

File: flaml/ml.py [hunk 1]
```diff
@@ -38,6 +38,35 @@
...
+                              "chrf": "min",
+                              "code_eval": "max",
+                              "comet": "max",
+                              "competition_math": "max",
+                              "coval": "max",
+                              "cuad": "max",
+                              "f1": "max",
+                              "gleu": "max",
+                              "google_bleu": "max",
+                             "matthews_correlation": "max",
```
Comment (sonichi) [inline flaml/ml.py:58]:
```suggestion
                              "google_bleu": "max",
                              "matthews_correlation": "max",

```

File: flaml/ml.py [hunk 2]
```diff
@@ -75,6 +104,51 @@ def get_estimator_class(task, estimator_name):
...
+                metric_mode = "max"
+            else:
+                metric = datasets.load_metric(metric_name)
+                metric_mode = huggingface_metric_to_mode[metric_name]
+                score = metric.compute(predictions=y_predict, references=y_true)[
+                                    metric_name
+                                ]
+            multiplier = -1 if metric_mode == "max" else 1
+            return score * multiplier
+        except FileNotFoundError:
```
Comment (sonichi) [inline flaml/ml.py:141]: Why expecting this error?

File: flaml/automl.py [hunk 3]
```diff
@@ -1996,8 +1992,16 @@ def custom_metric(
...
+                metric = load_default_huggingface_metric_for_task(self._state.task)
             else:
                 metric = "r2"
+
+        if _is_nlp_task(self._state.task):
+            self._state.fit_kwargs["metric"] = metric
+            self._state.fit_kwargs["use_ray"] = self._use_ray
+
         self._state.metric = metric
         if metric in [
```
Comment (sonichi) [inline flaml/automl.py:2006]: Update this part

**Review by sonichi — CHANGES_REQUESTED** [review]
I didn't comment on all the lint issues. Please fix them using pre-commit.

File: flaml/ml.py (hunk 2 above)
Comment (liususan091219) [reply flaml/ml.py:163]: Added comments

File: flaml/automl.py [hunk 4]
```diff
@@ -2003,6 +2004,10 @@ def custom_metric(
             self._state.fit_kwargs["use_ray"] = self._use_ray
 
         self._state.metric = metric
+
+        import datasets
```
Comment (sonichi) [inline flaml/automl.py:2008]: this will fail when [nlp] option is not installed.

File: flaml/automl.py [hunk 5]
```diff
@@ -2014,6 +2019,11 @@ def custom_metric(
             "micro_f1",
             "macro_f1",
             "ndcg",
+        ] + [
```
Comment (sonichi) [inline flaml/automl.py:2022]: ndcg@k is not included.

File: flaml/ml.py [hunk 6]
```diff
@@ -105,48 +122,62 @@ def get_estimator_class(task, estimator_name):
...
+                "currently built-in huggingface metrics are: "
+                + ", ".join(datasets.list_metrics())
+                + ", please pass a customized metric function to AutoML.fit(metric=func)"
+            )
+
+
+def is_in_sklearn_metric_name_set(metric_name):
+    if metric_name.startswith("ndcg") or metric_name in sklearn_metric_name_set:
+        return True
+    return False
```
Comment (sonichi) [inline flaml/ml.py:180]:
```suggestion
    return metric_name.startswith("ndcg") or metric_name in sklearn_metric_name_set
```

File: test/automl/test_multiclass.py [hunk 7]
```diff
@@ -366,7 +366,7 @@ def test_time_limit(self):
             learner_name="large_xgb", learner_class=MyLargeXGB
         )
         automl_settings = {
-            "time_budget": 0.5,
+            "time_budget": 5,
```
Comment (sonichi) [inline test/automl/test_multiclass.py:369]: Recover. This is a test for testing timeout.

Comment (sonichi) [reply test/automl/test_multiclass.py:369]: If line 378 fails, do a None check there.

File: flaml/ml.py [hunk 8]
```diff
@@ -162,14 +268,6 @@ def sklearn_metric_loss_score(
...
         else:
             score = 1 - ndcg_score([y_true], [y_predict])
-    else:
-        raise ValueError(
-            metric_name + " is not a built-in metric, "
-            "currently built-in metrics are: "
-            "r2, rmse, mae, mse, accuracy, roc_auc, roc_auc_ovr, roc_auc_ovo,"
-            "log_loss, mape, f1, micro_f1, macro_f1, ap. "
-            "please pass a customized metric function to AutoML.fit(metric=func)"
-        )
```
Comment (sonichi) [inline flaml/ml.py:172]: Recover and add "ndcg", "ndcg@k".

**Review by sonichi — CHANGES_REQUESTED** [review]

File: flaml/ml.py (hunk 8 above)
Comment (liususan091219) [reply flaml/ml.py:172]: I have it inside of the check for huggingface. 

Outside there's     
     if is_in_sklearn_metric_name_set(metric_name):
        return sklearn_metric_loss_score(
            metric_name, y_predict, y_true, labels, sample_weight, groups
        )

Which includes ndcg@k. So if the metric is outside of the sklearn metric, it won't enter sklearn_metric_loss_score.

File: flaml/automl.py [hunk 9]
```diff
@@ -1996,24 +1992,49 @@ def custom_metric(
...
+                if (
+                    metric in huggingface_metric_to_mode
+                    and huggingface_metric_to_mode[metric] == "max"
+                ):
+                    return True
+            return False
+
+        if isinstance(metric, str):
+            if is_to_reverse_metric(metric, task):
+                error_metric = f"1-{metric}"
```
Comment (sonichi) [inline flaml/automl.py:2035]: for hf metrics, it's "-metric" instead of "1-metric".

File: test/automl/test_multiclass.py [hunk 10]
```diff
@@ -375,13 +375,16 @@ def test_time_limit(self):
...
+        try:
+            print(automl_experiment.model.params)
+            automl_settings["estimator_list"] = ["large_xgb"]
+            automl_experiment.fit(X_train=X_train, y_train=y_train, **automl_settings)
+            print(automl_experiment.model)
+            automl_settings["estimator_list"] = ["large_lgbm"]
+            automl_experiment.fit(X_train=X_train, y_train=y_train, **automl_settings)
+            print(automl_experiment.model)
+        except AttributeError:
+            pass
```
Comment (sonichi) [inline test/automl/test_multiclass.py:387]: remove try...except and add a None check at line 379.

File: flaml/ml.py [hunk 11]
```diff
@@ -75,6 +121,63 @@ def get_estimator_class(task, estimator_name):
...
 
+def metric_loss_score(
+    metric_name,
+    y_predict,
+    y_true,
+    labels=None,
+    sample_weight=None,
+    groups=None,
+):
+    import datasets
```
Comment (sonichi) [inline flaml/ml.py:132]: this fails when [nlp] is not installed

**Review by sonichi — COMMENTED** [review]
I'm surprised "import datasets" didn't fail more tests.

File: flaml/ml.py (hunk 2 above)
Comment (sonichi) [reply flaml/ml.py:165]: if that exception only happens in load_metric(), the try...except needs to only contain that part.

**Review by sonichi — APPROVED** [review]

**Pull request merged**